src/segments/nba.go

```diff
@@ -0,0 +1,257 @@
+}
+
+func (nba *NBA) Enabled() bool {
+	data, err := nba.getResult()
+	if err != nil {
+		return false
+	}
+
+	nba.NBAData = *data
+
+	return true
+}
+
+func (nba *NBA) getResult() (*NBAData, error) {
+	// parses through a set of games and looks for props.team in away or home team
+	findGameByTeamTricode := func(games []Game, teamTricode string) (*Game, error) {
+		for _, game := range games {
+			if game.HomeTeam.TeamTricode == teamTricode || game.AwayTeam.TeamTricode == teamTricode {
+				return &game, nil
+			}
+		}
+		return nil, errors.New("no game found for team")
+	}
+	getCacheValue := func(key string) (*NBAData, error) {
+		val, found := nba.env.Cache().Get(key)
+		// we got something from the cache
+		if found {
+			var nbaData *NBAData
+			err := json.Unmarshal([]byte(val), &nbaData)
+			if err != nil {
+				return nil, err
+			}
+			return nbaData, nil
+		}
+		return nil, errors.New("no data in cache")
+	}
+
+	cachedScheduleKey := nba.props.GetString(TeamName, "") + "schedule"
+	cachedScoreKey := nba.props.GetString(TeamName, "") + "score"
+	httpTimeout := nba.props.GetInt(properties.HTTPTimeout, properties.DefaultHTTPTimeout)
+
+	// How often you want to query the API to get live score information, defaults to 2 minutes
+	cacheScoreTimeout := nba.props.GetInt(properties.CacheTimeout, 2)
+
+	// Cache the schedule information for a day so we don't call the API too often
+	cacheScheduleTimeout := nba.props.GetInt(properties.CacheTimeout, 14400)
+
+	teamName := nba.props.GetString(TeamName, "")
+
+	if cacheScheduleTimeout > 0 {
+		if data, err := getCacheValue(cachedScheduleKey); err == nil {
+			// check if the current time is after the start time of the game
+			// if so, we need to refresh the data
+			startTime, err := time.Parse("2006-01-02T15:04:05Z", data.StartTimeUTC)
+			if err != nil {
+				return nil, err
+			}
+			if time.Now().UTC().After(startTime) {
+				// remove the cache entry
+				nba.env.Cache().Delete(teamName)
+			} else {
+				return data, nil
+			}
+		}
+	}
+
+	if cacheScoreTimeout > 0 {
+		if data, err := getCacheValue(cachedScoreKey); err == nil {
+			return data, nil
+		}
+	}
+
+	body, err := nba.env.HTTPRequest(NBAScoreURL, nil, httpTimeout)
+	if err != nil {
+		// use oh my posh debugger
+		nba.env.Debug(fmt.Sprintf("Error: %s", err))
+		return nil, err
+	}
+	var scoreboardResponse *ScoreboardResponse
+	err = json.Unmarshal(body, &scoreboardResponse)
+	if err != nil {
+		nba.env.Debug(fmt.Sprintf("Error: %s", err))
+		return nil, err
+	}
+
+	gameInfo, err := findGameByTeamTricode(scoreboardResponse.Scoreboard.Games, teamName)
+	if err != nil {
+		nba.env.Debug(fmt.Sprintf("Error: %s", err))
+		return nil, err
+	}
+
+	data := NBAData{}
```

Comment:
any reason to use a second struct and not embed the `scoreboardResponse` in the NBA struct?

Reply:
Thanks for all the thoughtful feedback so far! I'll work on tweaking the other pieces based on your feedback :)

The reason for this was that the scoreboardResponse itself is quite large and can contain in theory (if every team is playing) 15 different game responses. Then in order to display the relevant information for the **specific team** you are hoping to follow you'd need to pull out the information for the specific team you are hoping to get schedule + stats for. Then inside of the NBA struct you'd have so much information for teams that are not relevant. There'd be a ton of extra info here and that to me felt a bit overwhelming to manage.

So the thought here was to parse the scoreboard response and then just cache the relevant information that we need for the team we are trying to get info for. I thought (not familiar with Go perf + considerations) this would be better for caching + performance. Two primary reasons:

1. If we store the whole scoreboardResponse in the cache we'd have a big chunk of data in the cache that we need to parse each time to get to the proper team info and most of that data may be irrelevant. We'd have to know where in the Games[] array the particular game is that we are trying to get information for.

2. If people want to create their own templates, they'd have a tough time because they'd have to know exactly where the data is for their team inside of the scoreboardResponse. With the NBAData struct it feels easy to just get any of the most relevant information that you'd actually care about.

That said ^, I'm completely new to the project, so lmk if you think there is a better way and I'm happy to learn from you and make the necessary changes :-)

Reply:
@jadoumie gotcha, but as we only care about what we will map towards the NBA struct, I would only create the relevant struct and properties instead of the entire JSON object. It's overhead that doesn't do anything except consume CPU cycles. So if this is what we need:

```go
data := NBAData{}
data.AwayTeam = gameInfo.AwayTeam.TeamTricode
data.HomeTeam = gameInfo.HomeTeam.TeamTricode
data.Time = gameInfo.GameStatusText
data.GameDate = scoreboardResponse.Scoreboard.GameDate
data.StartTimeUTC = gameInfo.GameTimeUTC
data.GameStatus = GameStatus(gameInfo.GameStatus)
data.HomeScore = gameInfo.HomeTeam.Score
data.AwayScore = gameInfo.AwayTeam.Score
data.HomeTeamWins = gameInfo.HomeTeam.Wins
data.HomeTeamLosses = gameInfo.HomeTeam.Losses
data.AwayTeamWins = gameInfo.AwayTeam.Wins
data.AwayTeamLosses = gameInfo.AwayTeam.Losses
```

Then this is all we need to parse:

```go
type Game struct {
	AwayTeam          Team        `json:"awayTeam"`
	HomeTeam          Team        `json:"homeTeam"`
	GameStatusText    string      `json:"gameStatusText"`
	GameTimeUTC       string      `json:"gameTimeUTC"`
	GameStatus        int         `json:"gameStatus"`
}

type Team struct {
	TeamTricode       string   `json:"teamTricode"`
	Score             int      `json:"score"`
	Wins              int      `json:"wins"`
	Losses            int      `json:"losses"`
}

type ScoreboardResponse struct {
	Scoreboard Scoreboard `json:"scoreboard"`
}

type Scoreboard struct {
	GameDate   string `json:"gameDate"`
}
```

Reply:
Oh wow! I didn't understand how this worked in go. I thought we needed to have a complete struct so we could map the byte array from the http request... makes sense that this would be way better though!

I'll go ahead and make the changes here!